### DISCUSSION TITLE
Gather up tags and format for YAML frontmatter

### DISCUSSION THREAD

**Discussion by luckman212** [SilentVoid13/Templater#140] [Templates Showcase]
Gathers all tags from the current file, sorts & de-duplicates them into a YAML-ready string for the frontmatter. This also lets you use the `#tagXXX...` format so you get the autocomplete, yet still end up with a clean taglist for YAML.

## Setup

Save these 2 templates in your templates folder. Assign a [hotkey](https://github.com/Vinzent03/obsidian-hotkeys-for-templates) for Template2 (I use <kbd>⌃ctrl ⇧shift T</kbd>). You can also assign a hotkey to Template1 if you want.

### Template1 (name it e.g. "YAML frontmatter tags")
```
---
tags: double_click_me
---

> Type tags below, then double-click the `double_click_me` in the YAML above and press <kbd>⌃ctrl ⇧shift T</kbd>.
>
> #<% tp.file.cursor() %>
>
> ==Delete this block when done.==
```

### Template2 (name this e.g. "tags from current file as YAML")

```
<%*
  function getUnique(value, index, self) {
    return self.indexOf(value) === index;
  }
  var out = '';
  var exclude_tags = [
    '#double_click_me'
  ];
  var tUnique = tp.file.tags.filter(getUnique);
  var tArr = tUnique.filter(t => exclude_tags.includes(t) === false);
  if (tArr.length) {
    tArr.sort();
    var tStr = tArr.join(', ').replace(/#/g,'');
    out = '[' + tStr + ']';
  }
%><%* tR += out %>
```

## Usage

Put your cursor near the top of your note, and then trigger **Template1**. Type your tags, then double-click where it says `double_click_me`

![image](https://user-images.githubusercontent.com/1992842/115766025-77b84c00-a375-11eb-8b28-be0f29085f02.png)

Then, trigger your hotkey for **Template2**.

_Result:_
![image](https://user-images.githubusercontent.com/1992842/115759487-4be59800-a36e-11eb-8d1b-bf1234de7716.png)

Lastly, put that YAML block at the top of your note and delete the temporary block.

_Final Result:_
![image](https://user-images.githubusercontent.com/1992842/115759656-7df6fa00-a36e-11eb-9daf-06118ebf6f04.png)

**Comment by lbiaggi** [comment]
Hi, btw thanks for sharing your Idea, I improved it a little bit, with this I don't need to be at the positions at tags and populate correctly, although it still doesn't update just insert.

```
<%*
  
  function getUnique(value, index, self) {
    return self.indexOf(value) === index;
  }
  var out = '';
  var exclude_tags = [
    '#double_click_me'
  ];
  // find the window
  let cmEditor =this.app.workspace.activeLeaf.view.editor
  /* I don't exactly how to find line direct, so I search around code mirror and 
	 found this which give offset from the begging of the file later I convert it to line, 
	 character position and increment it by 1 to
  */
  let tagspos = this.app.workspace.activeLeaf.view.data.search("tags:")
  let posLine = this.app.workspace.activeLeaf.view.editor.offsetToPos(tagspos)
  posLine.ch += (tagspos + 1)
  cmEditor.focus()
  cmEditor.setCursor(posLine)
  var tUnique = tp.file.tags.filter(getUnique);
  var tArr = tUnique.filter(t => exclude_tags.includes(t) === false);
  if (tArr.length) {
    tArr.sort();
    var tStr = tArr.join(', ').replace(/#/g,'');
    out = '[' + tStr + ']';
  }
%>
<%* tR += out %>
```

Regards,
Lucas
